File: 452. 删除链表中的元素.cpp
/**
 * Definition of singly-linked-list:
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *        this->val = val;
 *        this->next = NULL;
 *     }
 * }
 */

class Solution {
public:
    /**
     * @param head: a ListNode
     * @param val: An integer
     * @return: a ListNode
     */
    ListNode * removeElements(ListNode * head, int val) {
        if (head == nullptr) return nullptr;
        while (head != nullptr && head->val == val) {
            ListNode* del = head;
            head = head->next;
            delete del;
            del = nullptr;
        }
        if (head == nullptr) return nullptr;
        ListNode* prev = head;
        ListNode* curr = head->next;
        while (curr != nullptr) {
            if (curr->val == val) {
                ListNode* del = curr;
                curr = curr->next;
                prev->next = curr;
                delete del;
                del = nullptr;
            } else {
                prev = curr;
                curr = curr->next;
            }
        }
        return head;
    }
};

// 总耗时 399 ms
// 您的提交打败了 57.00% 的提交!

File: Python3/452. 删除链表中的元素.py
"""
Definition of ListNode
class ListNode(object):
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""

class Solution:
    """
    @param head: a ListNode
    @param val: An integer
    @return: a ListNode
    """
    def removeElements(self, head, val):
        if head == None :
            return head
        while head and head.val == val :
            head = head.next
        ref = head
        while ref and ref.next :
            if ref.next.val == val :
                ref.next = ref.next.next
            else :
                ref = ref.next
        return head

# 总耗时: 1611ms
# 您的提交打败了 58.71% 的提交!